[
   {
      "author_association" : "CONTRIBUTOR",
      "body" : "Concept ACK c1029c6ff3e58a64d0b07515f92938a717a42a02. This removes [database loading with no logs](https://www.youtube.com/watch?v=eUFY8Zw0Bag) behavior implemented in #2558 1859aafef0a273e27e886646e36140cc5e375ee1, which hopefully should be unnecessary now that the wallet flushes very frequently.\r\n\r\nThis change should reduce chances for data loss and provide better error feedback if there's an unsuccessful attempt to load newer bdb logs with an older version of bdb. It also simplifies the wallet opening code.\r\n\r\nIt's possible this PR could cause new errors for users who weren't seeing errors before. If this happens we could add an easier to use manual recovery option, or implement a middle-ground automatic approach like https://github.com/bitcoin/bitcoin/pull/18870#discussion_r421698704 that will retry loading the wallet without log files as long as the log files came from a compatible bdb version.",
      "created_at" : "2020-05-07T19:39:52Z",
      "html_url" : "https://github.com/bitcoin/bitcoin/pull/18907#issuecomment-625456981",
      "id" : 625456981,
      "issue_url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/18907",
      "node_id" : "MDEyOklzc3VlQ29tbWVudDYyNTQ1Njk4MQ==",
      "updated_at" : "2020-05-07T19:39:52Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/625456981",
      "user" : {
         "avatar_url" : "https://avatars2.githubusercontent.com/u/7133040?v=4",
         "events_url" : "https://api.github.com/users/ryanofsky/events{/privacy}",
         "followers_url" : "https://api.github.com/users/ryanofsky/followers",
         "following_url" : "https://api.github.com/users/ryanofsky/following{/other_user}",
         "gists_url" : "https://api.github.com/users/ryanofsky/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/ryanofsky",
         "id" : 7133040,
         "login" : "ryanofsky",
         "node_id" : "MDQ6VXNlcjcxMzMwNDA=",
         "organizations_url" : "https://api.github.com/users/ryanofsky/orgs",
         "received_events_url" : "https://api.github.com/users/ryanofsky/received_events",
         "repos_url" : "https://api.github.com/users/ryanofsky/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/ryanofsky/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/ryanofsky/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/ryanofsky"
      }
   },
   {
      "author_association" : "CONTRIBUTOR",
      "body" : "<!--e57a25ab6845829454e8d69fc972939a-->\n\nThe following sections might be updated with supplementary metadata relevant to reviewers and maintainers.\n\n<!--174a7506f384e20aa4161008e828411d-->\n### Conflicts\nReviewers, this pull request conflicts with the following ones:\n\n* #18971 (wallet: Refactor the classes in wallet/db.{cpp/h} by achow101)\n\nIf you consider this pull request important, please also help to review the conflicting pull requests. Ideally, start with the one that should be merged first.",
      "created_at" : "2020-05-14T03:49:06Z",
      "html_url" : "https://github.com/bitcoin/bitcoin/pull/18907#issuecomment-628369770",
      "id" : 628369770,
      "issue_url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/18907",
      "node_id" : "MDEyOklzc3VlQ29tbWVudDYyODM2OTc3MA==",
      "updated_at" : "2020-05-14T03:49:06Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/628369770",
      "user" : {
         "avatar_url" : "https://avatars2.githubusercontent.com/u/39886733?v=4",
         "events_url" : "https://api.github.com/users/DrahtBot/events{/privacy}",
         "followers_url" : "https://api.github.com/users/DrahtBot/followers",
         "following_url" : "https://api.github.com/users/DrahtBot/following{/other_user}",
         "gists_url" : "https://api.github.com/users/DrahtBot/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/DrahtBot",
         "id" : 39886733,
         "login" : "DrahtBot",
         "node_id" : "MDQ6VXNlcjM5ODg2NzMz",
         "organizations_url" : "https://api.github.com/users/DrahtBot/orgs",
         "received_events_url" : "https://api.github.com/users/DrahtBot/received_events",
         "repos_url" : "https://api.github.com/users/DrahtBot/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/DrahtBot/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/DrahtBot/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/DrahtBot"
      }
   },
   {
      "author_association" : "CONTRIBUTOR",
      "body" : "Some travis failures here. One in rpc_rawtransaction appears to be spurious https://travis-ci.org/github/bitcoin/bitcoin/jobs/684393890#L5036:\r\n\r\n```\r\n2020-05-07T18:54:10.222000Z TestFramework (ERROR): Unexpected exception caught during testing\r\nTraceback (most recent call last):\r\n  File \"/home/travis/build/bitcoin/bitcoin/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/test_framework/test_framework.py\", line 114, in main\r\n    self.run_test()\r\n  File \"/home/travis/build/bitcoin/bitcoin/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/rpc_rawtransaction.py\", line 299, in run_test\r\n    vout = next(o for o in rawTx['vout'] if o['value'] == Decimal('2.20000000'))\r\nStopIteration\r\n```\r\n\r\nOther two failures in wallet_tool test seem to be the result of changed error strings https://travis-ci.org/github/bitcoin/bitcoin/jobs/684393889#L3118 and https://travis-ci.org/github/bitcoin/bitcoin/jobs/684393892#L4350:\r\n\r\n```\r\nAssertionError: not(Error loading wallet.dat. Is wallet being used by other process? == Error initializing wallet database environment \"/home/travis/build/bitcoin/bitcoin/ci/scratch/test_runner/test_runner_â¿_ð_20200507_184719/tool_wallet_117/node0/regtest/wallets\"!\r\n```\r\n\r\n```\r\n  File \"/home/travis/build/bitcoin/bitcoin/ci/scratch/build/bitcoin-i686-pc-linux-gnu/test/functional/tool_wallet.py\", line 39, in assert_raises_tool_error\r\n    assert_equal(stderr.strip(), error)\r\n```",
      "created_at" : "2020-05-19T19:14:37Z",
      "html_url" : "https://github.com/bitcoin/bitcoin/pull/18907#issuecomment-631027177",
      "id" : 631027177,
      "issue_url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/18907",
      "node_id" : "MDEyOklzc3VlQ29tbWVudDYzMTAyNzE3Nw==",
      "updated_at" : "2020-05-19T19:15:40Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/631027177",
      "user" : {
         "avatar_url" : "https://avatars2.githubusercontent.com/u/7133040?v=4",
         "events_url" : "https://api.github.com/users/ryanofsky/events{/privacy}",
         "followers_url" : "https://api.github.com/users/ryanofsky/followers",
         "following_url" : "https://api.github.com/users/ryanofsky/following{/other_user}",
         "gists_url" : "https://api.github.com/users/ryanofsky/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/ryanofsky",
         "id" : 7133040,
         "login" : "ryanofsky",
         "node_id" : "MDQ6VXNlcjcxMzMwNDA=",
         "organizations_url" : "https://api.github.com/users/ryanofsky/orgs",
         "received_events_url" : "https://api.github.com/users/ryanofsky/received_events",
         "repos_url" : "https://api.github.com/users/ryanofsky/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/ryanofsky/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/ryanofsky/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/ryanofsky"
      }
   },
   {
      "author_association" : "MEMBER",
      "body" : "Cherrypicked the test and I think I've fixed the other test failures.",
      "created_at" : "2020-05-19T19:18:33Z",
      "html_url" : "https://github.com/bitcoin/bitcoin/pull/18907#issuecomment-631029099",
      "id" : 631029099,
      "issue_url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/18907",
      "node_id" : "MDEyOklzc3VlQ29tbWVudDYzMTAyOTA5OQ==",
      "updated_at" : "2020-05-19T19:18:33Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/631029099",
      "user" : {
         "avatar_url" : "https://avatars3.githubusercontent.com/u/3782274?v=4",
         "events_url" : "https://api.github.com/users/achow101/events{/privacy}",
         "followers_url" : "https://api.github.com/users/achow101/followers",
         "following_url" : "https://api.github.com/users/achow101/following{/other_user}",
         "gists_url" : "https://api.github.com/users/achow101/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/achow101",
         "id" : 3782274,
         "login" : "achow101",
         "node_id" : "MDQ6VXNlcjM3ODIyNzQ=",
         "organizations_url" : "https://api.github.com/users/achow101/orgs",
         "received_events_url" : "https://api.github.com/users/achow101/received_events",
         "repos_url" : "https://api.github.com/users/achow101/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/achow101/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/achow101/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/achow101"
      }
   },
   {
      "author_association" : "MEMBER",
      "body" : "> Travis is failing again currently because test_runner test list is o\r\n\r\nFixed",
      "created_at" : "2020-05-19T21:49:45Z",
      "html_url" : "https://github.com/bitcoin/bitcoin/pull/18907#issuecomment-631102297",
      "id" : 631102297,
      "issue_url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/18907",
      "node_id" : "MDEyOklzc3VlQ29tbWVudDYzMTEwMjI5Nw==",
      "updated_at" : "2020-05-19T21:49:45Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/631102297",
      "user" : {
         "avatar_url" : "https://avatars3.githubusercontent.com/u/3782274?v=4",
         "events_url" : "https://api.github.com/users/achow101/events{/privacy}",
         "followers_url" : "https://api.github.com/users/achow101/followers",
         "following_url" : "https://api.github.com/users/achow101/following{/other_user}",
         "gists_url" : "https://api.github.com/users/achow101/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/achow101",
         "id" : 3782274,
         "login" : "achow101",
         "node_id" : "MDQ6VXNlcjM3ODIyNzQ=",
         "organizations_url" : "https://api.github.com/users/achow101/orgs",
         "received_events_url" : "https://api.github.com/users/achow101/received_events",
         "repos_url" : "https://api.github.com/users/achow101/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/achow101/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/achow101/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/achow101"
      }
   }
]
